training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+- values: `uint(8)`
+  - the flattened array of bytes in all the strings (null byte delimited)
+- offsets: `int`
+  - the start index of each individual string in the `values`
+
+For the sake of simplicity, we'll treat values as a string array. So for our `s` above, these components look something like
+```
+values = ['s', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2', '\x00']
+offsets = [0, 5, 9]
+```
+For `getLengths`, we want to calculate the length of each individual string including the null terminator.
+
+```chapel
+const values: [0..#12] string = ['s', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2', '\x00'];
+const offsets = [0, 5, 9];
+const size = 3;
+
+/* Return lengths of all strings, including null terminator. */
+proc getLengths() {
+  // initialize lengths with the same domain as offsets
+  var lengths: [offsets.domain] int;
+  if size == 0 {
+    // if the segstring is empty, the lengths are also empty
+    return lengths;
+  }
+  // save off the last index of offsets
+  const high = offsets.domain.high;
+  forall (i, o, l) in zip(offsets.domain, offsets, lengths) {
+    if i == high {
+      // the last string
+      // len = values.size - start position of this string
+      l = values.size - o;
+    }
+    else {
+      // not the last string
+      // len = start poistion of next string - start position of this string
+      l = offsets[i+1] - o;
+    }
+  }
+  return lengths;
+}
+
+writeln(getLengths());
+```
+```console
+5 4 3
+```
+Nice! We'll use `getLengths`, `values`, and `offsets` in the next section as well.
+
+<a id="aggregation"></a>
+### Aggregation
+
+[Aggregation](https://chapel-lang.org/docs/modules/packages/CopyAggregation.html#copyaggregation) is used to copy
+local values into a remote array (`DstAggregator`) or copy values from a remote array into local variables (`SrcAggregator`).
+This provides a significant speed up when doing batch assignments.
+
+Syntax:
+```chapel
+// without aggregation
+forall (i, v) in zip(inds, vals) {
+  remoteArr[i] = v;
+}
+
+// with aggregation
+use CopyAggregation;
+forall (i, v) in zip(inds, vals) with (var agg = new DstAggregator(int)) {
+  agg.copy(remoteArr[i], v);
+}
+```
+It's important to note aggregation will only work if at least one side is local.
+Both sides being remote (remote-to-remote aggregation) is not yet supported. 
+
+<a id="ex16"></a>
+#### Example 16: Aggregation in Arkouda
+Let's do an example based on [`upper`](https://github.com/Bears-R-Us/arkouda/blob/b86adeb843275b0b86553534ede632acef4d15e2/src/SegmentedString.chpl#L427-L444) in `SegmentedString.chpl`.
+
+We will use `getLengths` and the `SegString s` from [example 15](#ex15):
+```chapel
+use CopyAggregation;
+
+/*
+  Given a SegString, return a new SegString with all lowercase characters from the original replaced with their uppercase equivalent
+  :returns: Strings – Substrings with lowercase characters replaced with uppercase equivalent
+*/
+proc upper() throws {
```

Comment:
Just noting that `throws` is used here without explanation or rationale.  (I think the rationale is that you're calling throwing functions and don't want to handle errors, so are throwing them upward to not have the compiler complain at you?  And that's true within Arkouda proper.  I _believe_ though that for a standalone toy example like this, you could probably leave off the `throws` and the program would halt if an error were thrown, complaining about it being unhandled.  The difference being whether you're within an explicit `module` declaration, or just sketching out some file-scope code like this (IIRC).

If I've got all that straight, I might just remove the `throws` from these examples.  If I'm wrong, you should probably give an explanation of why it's here. Or possibly introduce `throws` much earlier, like when `proc` is introduced?

Reply:
good point! I was following the reference function in arkouda a bit too closely here!